### ISSUE TITLE
[GenAI] Notebook fine-tuning LoRA/QLoRA — Qwen2.5-0.5B, 8 Go viable (lane po-2024:CoursIA, partition de #1454)

### ISSUE THREAD

**Issue by jsboige** [jsboige/CoursIA#10247]
## Pourquoi cette issue existe séparément de #1454

`po-2024` a **deux lanes** (`myia-po-2024:CoursIA` et `myia-po-2024:CoursIA-2`), et les deux ont demandé de la substance dans le même cycle. `#1454` (série RL/GRPO post-training, cadrage parcimonie RTX 3070) est **déjà dispatché sur la lane CoursIA-2** et y est le plat principal. Laisser les deux lanes converger sur `#1454` fabriquerait exactement la collision que `lane-claim-protocol` point 6 demande de partitionner à l'avance.

Cette issue est donc le grain **de la lane `myia-po-2024:CoursIA`** : famille GenAI, genre `genai` / `notebook-python`, aucun fichier en commun avec la série training.

## Scope

Ouvrir un notebook de **fine-tuning LoRA/QLoRA** dans la série GenAI : la mécanique du fine-tuning paramétriquement efficient, démontrée de bout en bout sur du matériel modeste.

Cadrage tranché (pas à re-litiger) :

- **Modèle jouet assumé** : Qwen2.5-0.5B en QLoRA 4-bit — viable sur 8 Go, et c'est le point pédagogique, pas une concession. On montre *le mécanisme* (rang, alpha, modules ciblés, ce qui est gelé et ce qui apprend), pas un SOTA de score.
- **Dataset petit et lisible** : quelques centaines d'exemples, un format d'instruction visible dans le notebook. Le lecteur doit pouvoir lire une paire prompt/réponse à l'œil.
- **La démonstration porte sur le delta** : sorties du modèle *avant* et *après* adaptation sur les mêmes prompts, côte à côte. Sans ce couple, le notebook n'a rien montré.
- **Le vrai outil** : `peft` + `transformers` + `bitsandbytes` réellement invoqués (Prong A, `sota-not-workaround`). Pas de simulation de courbe de perte, pas de sortie fabriquée. Si une pièce manque sur la machine, on l'installe (règle F) ; si elle est impossible, verdict SOTA écrit dans le body de la PR.

## Non-trivialité (Prong B)

Un LoRA sur un exemple où le modèle de base réussit déjà ne démontre rien — c'est le BFS-vs-A* du fine-tuning. Choisir une tâche où le base **échoue de façon visible** et où l'adaptation **corrige de façon visible** : format de sortie strict qu'il ne respecte pas, jargon de domaine absent, style systématiquement à côté. Et le **mesurer** avant de l'annoncer : la discrimination se constate, elle ne se plaide pas.

## Critères d'acceptation

- [ ] Notebook exécuté end-to-end, `execution_count != null` et outputs réels sur toute cellule code (C.2 / H.1)
- [ ] ≥ 3 exercices en stub (`pass` / `print("Exercice a completer")` / `return None` — **jamais** `raise NotImplementedError`, règle C.1)
- [ ] Couple avant/après sur ≥ 3 prompts identiques, dans les outputs committés
- [ ] Discrimination mesurée : une phrase qui dit ce que le base rate et ce que l'adapté réussit, appuyée sur les sorties du notebook
- [ ] Documentation primaire en français
- [ ] VRAM et wall-clock réels notés (machine + GPU nommés) — un lecteur sur 8 Go doit savoir à quoi s'attendre
- [ ] Verdict SOTA écrit dans le body de la PR (SOTA-OK attendu ; sinon `RECOVERABLE-*` avec ce qui a été tenté)

## Notes

- **Pas de dépendance au GPU-2 d'ai-01.** Le cadrage 0.5B QLoRA est précisément là pour que la lane soit autonome. Le slot 2B de `#5105` PR3 reste hors de cette issue.
- Le fantôme VRAM ~20 Go sur GPU1 ne gate pas ce grain.
- Série GenAI : vérifier l'emplacement et le README de série avant d'écrire (pas de doublon, mise à jour plutôt que création).

Part of #1385.

**Comment by jsboige** [comment]
[CLAIMED] lane myia-po-2024:CoursIA -- 2026-08-10T01:10Z (UTC). Grain DEEP/genai provisionné pour cette lane (option B' de mon provision-request c.76). Cadrage accepté, pas à re-litiger. Démarre ce cycle (worktree + structure notebook + premier draft). See #1454 (parent training/post-training).

**Comment by jsboige** [comment]
[DELIVERED] lane myia-po-2024:CoursIA — résolu par PR #10328 (merged 2026-08-10T17:19Z).

Vérification firsthand de `21_LoRA_FineTuning.ipynb` sur `origin/main` contre les 7 critères d'acceptation :

- [x] **Notebook exécuté end-to-end, `execution_count != null`, outputs réels** — 12 cellules code, **0 null-exec**.
- [x] **≥ 3 exercices en stub** — 3 cellules stub (`pass` / `# TODO`), **aucun** `raise NotImplementedError` (C.1 conforme).
- [x] **Couple avant/après sur prompts identiques** — sections « avant » et « après » présentes dans les outputs committés.
- [x] **Discrimination mesurée** — le delta Prong-B (format de sortie strict que le base rate et l'adapté corrige) est documenté et reproduit sur l'env courant (le corps de #10328 détaille la re-productibilité du delta sous trl 1.9.2).
- [x] **Documentation primaire en français**.
- [x] **VRAM + wall-clock réels notés (machine + GPU nommés)** — VRAM + Go présents ; le cadre 0.5B QLoRA est précisément le point pédagogique 8 Go viable.
- [x] **Verdict SOTA écrit dans le body de la PR** — #10328 porte le tag `Grain: MED/notebook-python — lane myia-po-2024:CoursIA` et documente les deux défauts de reproductibilité (API trl 0.16.1 → 1.9.2 + delta Prong-B) firsthand-vérifiés sur `coursia-ml-training`.

Outil réel invoqué (Prong A) : `peft` + `transformers` + `bitsandbytes` (QLoRA 4-bit), pas de simulation.

Tous les critères d'acceptation remplis. Je laisse ai-01 confirmer/fermer (issue ouverte par le coordinateur). See #1454 (série RL/GRPO), #5105 (slot 2B hors scope).

**Comment by jsboige** [comment]
[CLAIMED] lane myia-po-2024:CoursIA — re-cible LoRA/QLoRA sur Qwen3.5-0.8B-Base (mandat user, steer ai-01 msg-20260810T183412-mndjfo). Notebook 21_LoRA_FineTuning passe de Qwen2.5-0.5B (toy) à Qwen3.5-0.8B (SOTA petit modèle). QLoRA 4-bit sur RTX 3070 8GB. Verdict modèle mesuré (VRAM/erreur) si 3.5 ne passe pas. Tente 3.5-2B en quant si VRAM le permet.

**Comment by jsboige** [comment]
[DELIVERED] lane myia-po-2024:CoursIA — re-cible Qwen3.5-0.8B-Base RESOLUE par PR #10361 (merged, commit e3a8de82b2).

Firsthand verification de `21_LoRA_FineTuning.ipynb` sur origin/main (G.9, pas sur le status condense) — les 7 criteres d acceptation tous satisfaits :

- [x] **Execute end-to-end, ec!=null, outputs reels** — 12 cellules code, ec 1..12 contigus, 0 null-exec.
- [x] **>=3 exercices en stub** — 3 stubs (`pass`/`# TODO`), aucun `raise NotImplementedError` (C.1).
- [x] **Couple avant/apres sur >=3 prompts** — cell17 : Sharpe ratio / Gradient descent / Overfitting.
- [x] **Discrimination mesuree (delta 0/3 -> 3/3)** — cell18 firsthand :
  ```
  Sharpe ratio      AVANT=False  APRES=True
  Gradient descent  AVANT=False  APRES=True
  Overfitting       AVANT=False  APRES=True
  ```
  Base Qwen3.5-0.8B genere du HTML (`<b>`) ; apres LoRA genere le format etranger ⟦T⟧..⟦/T⟧⟦D⟧..⟦/D⟧⟦E⟧..⟦/E⟧. Format DELIBEREMENT foreign au base (lora-prongb-delimiter-must-be-foreign-to-base).
- [x] **Docs FR** — documentation primaire en francais.
- [x] **VRAM notee** — 0.78 Go au chargement, 2.51 Go pic pendant training / 8 Go (RTX 3070 Laptop).
- [x] **Wall-clock note** — 5.7s chargement, 1593.6s training, perte finale 0.6147.

Note : `Tente 3.5-2B en quant si VRAM le permet` (de mon claim) — non tente ; 0.8B SOTA petit modele suffit a la discrimination, 2B = enhancement future (non bloquant, tracker a part si souhaite).

Issue entierement resolue (3 PRs : #10251 original Qwen2.5-0.5B, #10328 trl1.9.2 re-exec, #10361 re-cible Qwen3.5-0.8B). Fermeture.

See #1454 (parent training/post-training, reste ouvert).
